themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,193 @@ function toggleVisibility(e) {
+	if (active && search) {
+		currentPageSearch(search);
+		menuItemsSearch(search);
+	}
+}
+
+function searchOnHashChange() {
+	const search = getSearch();
+	updateMenuLinks(search);
+	currentPageSearch(search);
+	menuItemsSearch(search);
+}
+
+function currentPageSearch(search) {
+	const h2 = document.getElementsByTagName('h2');
+	const p = document.getElementsByTagName('p');
+	let scrolled = false;
+	for (let searchTarget of [...h2, ...p]) {
+		const a = searchTarget.getElementsByTagName('a')[0];
+		if (a && a.textContent) {
+			searchTarget = a;
+		}
+		if (highlightText(searchTarget, search) && !scrolled) {
+			scrolled = true;
+			searchTarget.scrollIntoView();
+		}
+	}
+}
+
+function menuItemsSearch(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	const titles = nav.querySelectorAll('.title');
+	for (let title of titles) {
+		const menuItem = title.parentElement;
+		const tags = menuItem.dataset.tag || '';
+		for (let tag of tags.split(',')) {
+			if (tag.toLowerCase().indexOf(search) >= 0 && search != '') {
+				const matchingTag = findMatchingTag(tag, search);
+				showTag(menuItem, tag, matchingTag);
+				const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0].querySelector('a');
+				highlightText(dataShowTag, search);
+				break;
 			}
+			hideTag(menuItem, search);
+		}
+		highlightText(title, search)
+	}
+}
+
+function findMatchingTag(tag, search) {
+	for (let searchTarget of tag.split(' ')) {
+		let indexTarget = -1;
+		for (let arraySearch of search.split(' ')) {
+			if (indexTarget < 0 && arraySearch != '') {
+				indexTarget = searchTarget.toLowerCase().indexOf(arraySearch);
+				break;
+			}
+		}
+		const indexSearch = search.split(' ').indexOf(searchTarget.toLowerCase());
+		if ((indexTarget >= 0 || indexSearch >= 0 || search == ' ') && search != '') {
+			return searchTarget;
+		}
+	}
+}
+
+function showTag(menuItem, tag, searchTarget) {
+	if (menuItem.getElementsByTagName('data-show-tag').length === 0) {
+		createDataShowTag(menuItem);
+	}
+	focusOnTag(menuItem);
+	const a = menuItem.getElementsByTagName('data-show-tag')[0].firstChild;
+	a.textContent = getFormattedTagsContent(tag, searchTarget);
+	a.setAttribute('href', menuItem.querySelector('.title').getAttribute('href'));
+	a.style.display = 'inline';
+}
+
+function hideTag(menuItem, search) {
+	const title = menuItem.querySelector('.title');
+	menuItem.style.display = (search == '') ? '' : 'none';
+	const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0];
+	if (dataShowTag) {
+		dataShowTag.style.display = 'none';
+	}
+	title.classList.remove('menu-item');
+	title.textContent = title.textContent.replace("/ ", '');
+}
+
+function createDataShowTag(menuItem) {
```

Comment:
don't need to create a separate function for this
showTag and craeteDataShowTag looks the same